src/mgr/TTLCache.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <boost/optional.hpp>
+#include <chrono>
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <string>
+
+using namespace std;
+
+template <class Key, class Value> class Cache {
+private:
+  uint64_t hits = 0;
+  uint64_t misses = 0;
+  std::mutex hit_mutex, miss_mutex;
+
+protected:
+  unsigned int capacity;
+  Cache(unsigned int size = UINT16_MAX) : capacity{size} {};
+  std::map<Key, Value> content;
+
+  void miss() {
+    miss_mutex.lock();
+    misses++;
+    miss_mutex.unlock();
+  }
+  void hit() {
+    hit_mutex.lock();
+    hits++;
+    hit_mutex.unlock();
+  }
+
+  unsigned int get_misses() { return misses; }
+  unsigned int get_hits() { return hits; }
+
+public:
+  void insert(Key key, Value value) {
+    if (content.size() < capacity) {
+      content.insert({key, value});
+    }
+  }
+  Value get(Key key) { return content[key]; }
+  void erase(Key key) { content.erase(content.find(key)); }
+  void clear() { content.clear(); }
+  bool exists(Key key) { return content.find(key) != content.end(); }
+  std::pair<uint64_t, uint64_t> get_hit_miss_ratio() {
+    return std::make_pair(hits, misses);
+  }
+  int size() { return content.size(); }
+  ~Cache(){};
+};
+
+typedef std::chrono::time_point<std::chrono::steady_clock> ttl_time_point;
+template <class Key, class Value>
+class TTLCache : public Cache<Key, std::pair<Value, ttl_time_point>> {
+private:
+  uint16_t ttl;
+  typedef std::pair<Value, ttl_time_point> value_type;
+
+  Value get_value(Key key);
+  ttl_time_point get_value_time_point(Key key);
+  bool exists(Key key);
+  bool expired(Key key);
+  std::function<void(Value)> erase_callback = nullptr;
+
+public:
+  TTLCache(uint16_t ttl_ = 0, uint16_t size = UINT16_MAX)
+      : Cache<Key, value_type>(size), ttl{ttl_} {
+
+  }
+  ~TTLCache(){};
+  void insert(Key key, Value value);
+  boost::optional<Value> get(Key key);
+  void erase(Key key);
+  void clear();
+  void set_ttl(uint16_t ttl);
+  void set_erase_callback(std::function<void(Value)> cb) { erase_callback = cb; }
+};
+
+template <class Key, class Value>
+void TTLCache<Key, Value>::insert(Key key, Value value) {
+  auto now = std::chrono::steady_clock::now();
+
+  if (!ttl)
+    return;
+  uint16_t ttl_spread = ttl * 0.25;
```

Comment:
I'd rather expose the ratio and make it default to 0.25 rather than hard-coding it. It's a setting potentially subject to fine-tuning during your optimization stage.